src/Microsoft.CodeAnalysis.BannedApiAnalyzers/Core/RestrictedInternalsVisibleToAnalyzer.cs

```diff
@@ -0,0 +1,237 @@
+
+            context.RegisterCompilationStartAction(OnCompilationStart);
+        }
+
+        private void OnCompilationStart(CompilationStartAnalysisContext compilationContext)
+        {
+            var restrictedInternalsVisibleToMap = GetRestrictedInternalsVisibleToMap(compilationContext.Compilation);
+            if (restrictedInternalsVisibleToMap.IsEmpty)
+            {
+                return;
+            }
+
+            var namespaceToIsBannedMap = new ConcurrentDictionary<INamespaceSymbol, /*isBanned*/bool>();
+
+            // Verify all explicit type name specifications in declarations and executable code.
+            compilationContext.RegisterSyntaxNodeAction(
+                context =>
+                {
+                    var name = (TNameSyntax)context.Node;
+                    if (!IsInTypeOnlyContext(name) ||
+                        name.Parent is TNameSyntax)
+                    {
+                        // Bail out if we are not in type only context or the parent is also a name
+                        // which will be analyzed separately.
+                        return;
+                    }
+
+                    var typeInfo = context.SemanticModel.GetTypeInfo(name, context.CancellationToken);
+                    VerifySymbol(typeInfo.Type as INamedTypeSymbol, name, context.ReportDiagnostic);
+                },
+                NameSyntaxKinds);
+
+            // Verify all member usages in executable code.
+            compilationContext.RegisterOperationAction(
+                context =>
+                {
+                    ISymbol symbol;
+                    switch (context.Operation)
+                    {
+                        case IObjectCreationOperation objectCreation:
+                            symbol = objectCreation.Constructor;
+                            break;
+                        case IInvocationOperation invocation:
+                            symbol = invocation.TargetMethod;
+                            break;
+                        case IMemberReferenceOperation memberReference:
+                            symbol = memberReference.Member;
+                            break;
+                        default:
+                            throw new NotImplementedException($"Unhandled OperationKind: {context.Operation.Kind}");
+                    }
+
+                    VerifySymbol(symbol, context.Operation.Syntax, context.ReportDiagnostic);
+                },
+                OperationKind.ObjectCreation,
+                OperationKind.Invocation,
+                OperationKind.EventReference,
+                OperationKind.FieldReference,
+                OperationKind.MethodReference,
+                OperationKind.PropertyReference);
+
+            return;
+
+            // Local functions.
+            static ImmutableDictionary<IAssemblySymbol, ImmutableSortedSet<string>> GetRestrictedInternalsVisibleToMap(Compilation compilation)
```

Comment:
It's be nice if we can cache this data somehow, given that for consumers, the assembly reference doesn't change much.

Reply:
Yes, but that would need to live across different compilations. We can come back to this performance optimization if we require it in future.